lib/elastic_apm/central_config.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/central_config/cache_control'
+
+module ElasticAPM
+  # @api private
+  class CentralConfig
+    include Logging
+
+    # @api private
+    class ResponseError < InternalError
+      def initialize(response)
+        @response = response
+      end
+
+      attr_reader :response
+    end
+    class ClientError < ResponseError; end
+    class ServerError < ResponseError; end
+
+    DEFAULT_MAX_AGE = 300
+
+    def initialize(config)
+      @config = config
+      @service_info = {
+        'service.name': config.service_name,
+        'service.environment': config.environment
+      }.to_json
+    end
+
+    attr_reader :config, :task
+
+    def start
+      fetch_and_apply_config
+    end
+
+    def fetch_and_apply_config
+      @task&.cancel
+      @task =
+        Concurrent::Promise
+        .execute(&method(:fetch_config))
+        .on_success(&method(:handle_success))
+        .rescue(&method(:handle_error))
+    end
+
+    def stop
+      @task&.cancel
+    end
+
+    # rubocop:disable Metrics/MethodLength
+    def fetch_config
+      resp = perform_request
+
+      case resp.status
+      when 200..299
+        resp
+      when 300..399
+        resp
+      when 400..499
+        raise ClientError, resp
+      when 500..599
+        raise ServerError, resp
+      end
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    private
+
+    def handle_success(resp)
+      unless resp.body.empty?
+        update = JSON.parse(resp.body)
+        config.assign(update)
```

Comment:
It looks like you're just splatting the new config over the top, is that right? What happens when the config is removed in Kibana? We should fall back to locally defined config (https://github.com/elastic/apm/issues/76#issuecomment-510019654). If you're deferring that, please document the limitation and create a TODO/issue.

Reply:
Got it. Nice catch.